File: one-more-time/Lagrange/Lagrange.fs
open System
open System.IO

type Point = { X: float; Y: float }

let zeros n =
    Array.zeroCreate n

let denom i points =
    let x = (points: Point []).[i].X
    let mutable p = 1.0
    for j = Array.length points - 1 downto 0 do
        if i <> j then
            p <- p * (x - points.[j].X)
    p

let interpolate i points =
    let length = Array.length points
    let coeff = zeros length
    coeff.[0] <- 1.0 / denom i points

    for k = 0 to Array.length points - 1 do
        if k <> i then
            let newCoeffArray = zeros length
            for j = (if k < i then k + 1 else k) downto 0 do
                if j < 2 then
                    newCoeffArray.[j + 1] <- newCoeffArray.[j + 1] + coeff.[j]
                    newCoeffArray.[j] <- newCoeffArray.[j] - points.[k].X * coeff.[j]
                    //printf "%d " k
                    //printfn "%d" j
                    //printf "%.2f " newCoeffArray.[0]
                    //printf "%.2f " newCoeffArray.[1]
                    //printfn "%.2f" newCoeffArray.[2]
            for i = 0 to Array.length points - 1 do
                coeff.[i] <- newCoeffArray.[i]
            coeff
        else
            coeff

    coeff

let lagrange points =
    let length = Array.length points
    let polynom = zeros length

    let rec loop i =
        if i < length then
            let coeff = interpolate i points
            for k = 0 to length - 1 do
                polynom.[length - 1 - k] <- polynom.[length - 1 - k] + points.[i].Y * coeff.[k]
            loop (i + 1)
    
    loop 0
    polynom

let main() =
    let lines = File.ReadAllLines("ingput.txt")
    let degree = int lines.[0]
    let points = Array.init (degree + 1) (fun i ->
        let info = lines.[i + 1].Split()
        let x = Double.Parse info.[0]
        let y = Double.Parse info.[1]
        { X = x; Y = y })

    let finalCoeff = lagrange points
    use writer = File.CreateText("otput.txt")
    finalCoeff |> Array.iter (fun coeff -> writer.Write("{0} ", coeff))
    //for i = 0 to finalCoeff.Length-1 do  
        //printfn "%.2f" finalCoeff.[i]
    printfn "Lagrange coefficients successfully written."

main()

File: one-more-time/Lagrange/Lagrange.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Arrays;

public class Lagrange {
    public static double[] zeros(int n) {
        double[] arr = new double[n];
        Arrays.fill(arr, 0);
        return arr;
    }

    static class Point {
        double x, y;
    
        Point(double x, double y) {
            this.x = x;
            this.y = y;
        }
    }

    public static double denom(int i, Point[] points) {
        double p = 1;
        double x = points[i].x;
        for (int j = 0; j < points.length; j++) {
            if (i != j) {
                p *= x - points[j].x;
            }
        }
        return p;
    }

    public static double[] interpolate(int i, Point[] points) {
        double[] coeff = zeros(points.length);
        coeff[0] = 1 / denom(i, points);

        for (int k = 0; k < points.length; k++) {
            if (k == i) {
                continue;
            }
            double[] newCoeff = zeros(points.length);
            for (int j = (k < i) ? k + 1 : k; j >= 0; j--) {
                if (j < 2) {
                    newCoeff[j + 1] += coeff[j];
                    newCoeff[j] -= points[k].x * coeff[j];
                }
            }
            coeff = newCoeff;
        }
        return coeff;
    }

    public static double[] lagrange(Point[] points) {
        double[] polynom = zeros(points.length);
        double[] coeff;

        for (int i = 0; i < points.length; i++) {
            coeff = interpolate(i, points);
            for (int k = 0; k < points.length; k++) {
                polynom[points.length - 1 - k] += points[i].y * coeff[k];
            }
        }
        return polynom;
    }

    public static void main(String[] args) throws IOException {
        Point[] points = null;
        try (BufferedReader br = new BufferedReader(new FileReader("ingput.txt"))) {
            int degree = Integer.parseInt(br.readLine());
            points = new Point[degree + 1];
            for (int i = 0; i < degree + 1; i++) {
                String[] info = br.readLine().split(" ");
                double x = Double.parseDouble(info[0]);
                double y = Double.parseDouble(info[1]);
                points[i] = new Point(x, y);
            }
        }

        double[] finalCoeff = lagrange(points);
        StringBuilder str = new StringBuilder();
        for (double coeff : finalCoeff) {
            str.append(coeff).append(" ");
        }

        try (FileWriter writer = new FileWriter("otput.txt")) {
            writer.write(str.toString());
            System.out.println("Lagrange coefficients successfully written.");
        }
    }
}
